src/logic/map_objects/world/critter.cc

```diff
@@ -230,25 +277,192 @@ Simply roam the map
+			assert(!foundme);
+			foundme = true;
+			continue;
+		}
+		if (!foundme) {
+			++n_th_bob_on_field;
+		}
+		if (descr().name() == b->descr().name()) {
+			++mating_partners;
+		}
+		if (upcast(Critter, c, b)) {
+			other_herbivores_on_field |= c->descr().is_herbivore();
+		}
+	}
+	assert(foundme);
+	if (descr().is_carnivore()) {
+		// only hunt other carnivores if there are no herbivores here
+		for (Bob* b = get_position().field->get_first_bob(); b; b = b->get_next_bob())
+			if (descr().name() != b->descr().name())
+				if (upcast(Critter, c, b))
+					if (!c->descr().is_carnivore() || !other_herbivores_on_field)
+						candidates_for_eating.push_back(c);
+	}
+	while (!candidates_for_eating.empty()) {
+		size_t idx = game.logic_rand() % candidates_for_eating.size();
+		if (game.logic_rand() % 100 < descr().get_appetite()) {
+			// yum yum yum
+			Critter* food = candidates_for_eating[idx];
+			bool skipped = false;
+			if (food) {
+				molog("Yummy, I love a %s...\n", food->descr().name().c_str());
+				// find hunting partners
+				int32_t attacker_strength = 0;
+				int32_t defender_strength = 0;
+				if (food->descr().is_carnivore()) {
+					// 1vs1
+					attacker_strength = descr().get_size();
+					defender_strength = food->descr().get_size();
+				} else {
+					// ffa
+					for (Bob* b = get_position().field->get_first_bob(); b; b = b->get_next_bob()) {
+						if (upcast(Critter, c, b)) {
+							if (c != this && (c == food || c->descr().is_herbivore())) {
+								defender_strength += c->descr().get_size();
+							} else if (c->descr().is_carnivore()) {
+								attacker_strength += c->descr().get_size();
+							}
+						}
+					}
+				}
+				assert(attacker_strength > 0);
+				assert(defender_strength > 0);
+				const int32_t S =
+				   attacker_strength - defender_strength;  // weighted combined strength *difference*
+				const int32_t N = defender_strength * defender_strength;  // definition of N: the
+				                                                          // resulting chance is 1 if
+				                                                          // and only if S >= N
+				const double weighted_success_chance =
+				   S <= -N ? 0.0 : S >= N ? 1.0 :
+				                            -(std::log(S * S + 1) - 2 * S * std::atan(S) - M_PI * S -
+				                              std::log(N * N + 1) + N * (2 * std::atan(N) - M_PI)) /
+				                               (2 * N * M_PI);
+				molog("    *** [total strength %d] vs [prey %d] *** success chance %lf\n", S,
+				      defender_strength, weighted_success_chance);
+				assert(weighted_success_chance >= 0.0);
+				assert(weighted_success_chance <= 1.0);
+				if (game.logic_rand() % N < weighted_success_chance * N) {
+					molog("    SUCCESS :)\n");
+					food->remove(game);
+				} else {
+					molog("    failed :(\n");
+					skipped = true;
+				}
+			} else {
+				// refers to the plant on our field
+				upcast(Immovable, imm, get_position().field->get_immovable());
```

Comment:
We have this loop above that could already store the immovable into a separate variable rather than hacking the Critter* vector with a nullptr

```
	if (descr().is_herbivore() && get_position().field->get_immovable()) {
		for (uint32_t a : descr().food_plants()) {
			if (get_position().field->get_immovable()->descr().has_attribute(a)) {
				candidates_for_eating.push_back(nullptr);
				break;
			}
		}
	}
```

Reply:
It is not currently used, but theoretically we support omnivores. If an omnivore enters a field with a plant and some animals on it, all of these should have equal chances to be picked as dinner. That's why I'd prefer to have only one list for all food items here.

Reply:
You could run the function that calculates the chance on `size() + 1` and pick the immovable if the `result == size()`. That's cheaper than having multiple loops with dynamic casts.

Reply:
The point of having a `while` loop here is that if eating the first prey doesn't succeed, we try again to get another meal, until it succeeds or every prey escaped.
We no longer have any casts in this loop except one `upcast(Immovable, imm, get_position().field->get_immovable())` which is performed only when we are 100% certain that we want to eat the upcasted plant…

Reply:
Of course, the `while (!candidates_for_eating.empty()) {` needs to stay. The repetition I wanted to get rid of was multiple walks through `for (Bob* b = get_position().field->get_first_bob(); b; b = b->get_next_bob()) {`.

The immovable will always he accessed first though if any, so I'd still prefer to not have the nullptr hack. `candidates_for_eating.push_back(nullptr);` Could be replaced by storing the immovable into a variable, and the check can then be whether that variable is nullptr.

Reply:
Which `for (Bob* b = get_position().field->get_first_bob(); b; b = b->get_next_bob())` exactly are you referring to?
And
> The immovable will always he accessed first though if any

If so that would be a bug. In each iteration we pick a random index in `candidates_for_eating` so this does not necessarily come first

Reply:
You can fix that problem by using

```
size_t idx = game.logic_rand() % (candidates_for_eating.size() + 1);
if (idx == candidates_for_eating.size()) {
  // pick immovable
} else {
  // pick from candidates_for_eating
}
```

Reply:
:)